dev/doc/parsing.md

```diff
@@ -0,0 +1,378 @@
+
+* The grammar is split into multiple source files.  Nonterminals can be local to a file
+  or global.
+
+* While 95% of the nonterminals and almost all the productions are defined in the grammar,
+  a few are defined directly in OCaml code.  Since many developers have worked on the parser
+  over the years, this code can be idiosyncratic, reflecting various design concepts.
+
+* The parser is a recursive descent parser that, by default, only looks at the next token
+  to make a parsing decision.  It's possible to hand-code additional lookahead where
+  necessary by writing OCaml code.
+
+* There's no code that checks whether a grammar is ambiguous or whether every production
+  is recognized.  Developers who modify the grammar may, in some cases, need to structure their
+  added productions in specific ways to ensure that their additions are parsable and that they
+  don't break existing productions.
+
+## Contents ##
+
+- [Grammars: `*.mlg` File Structure](#grammars-mlg-file-structure)
+- [Grammars: Nonterminals and Productions](#grammars-nonterminals-and-productions)
+  - [Alternate production syntax](#alternate production-syntax)
+- [Usage notes](#usage-notes)
+  - [Other components](#other-components)
+  - [Parsing productions](#parsing-productions)
+  - [Lookahead](#lookahead)
+
+## Grammars: `*.mlg` File Structure ##
+
+Grammars are defined in `*.mlg` files, which `coqpp` compiles into `*.ml` files at build time.
+`coqpp` code is in `coqpp` directory.  `coqpp` uses yacc and lex to parse the grammar files.
+You can examine its yacc and lex input files in `coqpp_lex.mll` and `coqpp_parse.mly` for
+details not fully covered here.
+
+In addition, there is a `doc_grammar` build utility that uses the `coqpp` parser to extract the
+grammar, then edits and inserts it into the documentation.  This is described in `doc/tools/docgram/README.md`.
+`doc_grammar` generates `doc/tools/docgram/fullGrammar`, which has the full grammar for Coq
+(not including some optionally-loaded plugins).  This may be easier to read since everything is
+in one file and the parser action routines and other OCaml code are omitted.
+
+`*.mlg` files contain the following types of nodes (See `node` in the yacc grammar).  This part is
+very specific to Coq (not so similar to Camlp5):
+
+* OCaml code - OCaml code enclosed in curly braces, which is copied verbatim to the generated `*.ml` file
+
+* Comments - comments in the `*.mlg` file in the form `(* ... *)`, which are dropped.  Comments
+  in OCaml code are preserved.
+
+* `DECLARE_PLUGIN "*_plugin"` - associates the file with a specific plugin, for example "ltac_plugin"
+
+* `GRAMMAR EXTEND` - adds additional nonterminals and productions to the grammar and declares global
+  nonterminals referenced in the `GRAMMAR EXTEND`:
+
+  ```
+  GRAMMAR EXTEND Gram
+    GLOBAL:
+      bignat bigint ...;
+    <nonterminal definitions>
+  END
+  ```
+
+  TODO: What is `Gram` above?  Worth explaining?
```

Comment:
Nowadays it is dead code left there only for backwards compatibility. It used to be the OCaml name of the grammar being extended, but now coqpp calls another primitive.